domain/application/resource/store.go

```diff
@@ -4,64 +4,96 @@
 package resource
 
 import (
-	"context"
-	"fmt"
-	"io"
+    "context"
+    "fmt"
+    "io"
+    "path"
 
-	"github.com/juju/juju/core/objectstore"
-	"github.com/juju/juju/internal/charm/resource"
+    "github.com/juju/juju/core/application"
+    "github.com/juju/juju/core/objectstore"
+    "github.com/juju/juju/internal/charm/resource"
 )
 
 // ResourceStore provides a list of methods necessary for interacting with
 // a store for the resource.
 type ResourceStore interface {
-	// Get returns an io.ReadCloser for data at path, namespaced to the
-	// model.
-	Get(context.Context, string) (io.ReadCloser, int64, error)
+    // Get returns an io.ReadCloser for a resource in the resource store.
+    Get(ctx context.Context, applicationUUID application.ID, name string) (r io.ReadCloser, size int64, err error)
+
+    // Put stores data from io.Reader in the resource store at the
+    // path specified in the resource.
+    Put(ctx context.Context, r io.Reader, resource Resource) error
+
+    // Remove removes a resource from storage.
+    Remove(ctx context.Context, applicationUUID application.ID, name string) error
 }
 
 // FileResourceStore is a ResourceStore for file resource types.
 type FileResourceStore struct {
-	objectStore objectstore.ObjectStore
+    objectStore objectstore.ObjectStore
+}
+
+// Get the specified resource from the object store.
+func (f FileResourceStore) Get(
+    ctx context.Context,
+    applicationUUID application.ID,
+    name string,
+) (io.ReadCloser, int64, error) {
+    path := f.storagePath(name, applicationUUID)
+    return f.objectStore.Get(ctx, path)
+}
+
+// Put the given resource in the object store.
+func (f FileResourceStore) Put(ctx context.Context, r io.Reader, resource Resource) error {
+    path := f.storagePath(resource.Name, resource.ApplicationID)
+    return f.objectStore.PutAndCheckHash(ctx, path, r, resource.Size, resource.Fingerprint.String())
+}
+
+// Remove the specified resource from the object store.
+func (f FileResourceStore) Remove(ctx context.Context, applicationUUID application.ID, name string) error {
+    path := f.storagePath(name, applicationUUID)
+    return f.objectStore.Remove(ctx, path)
 }
 
-func (f FileResourceStore) Get(ctx context.Context, path string) (io.ReadCloser, int64, error) {
-	return f.objectStore.Get(ctx, path)
+// storagePath returns the path at which the resource will be stored in the
+// object store. It is stored under the application name.
```

Comment:
**todo:** I think you meant "application uuid"

Reply:
Very true.